Payment Request button:  Various Improvements

asumaran opened the pull request:
# Changes proposed in this Pull Request:

Part of #1563. #1587 follow-up. 

Minor improvements.

<!--
Enables the Payment Request button on the cart page for subscriptions created using the [WooCommerce Subscriptions plugin](https://woocommerce.com/products/woocommerce-subscriptions/), virtual subscriptions, downloadable and variable subscriptions. **It doesn't work for Trial Subscriptions that require Shipping** though. Will create a separate issue for it. 

# Testing instructions

First of all you need to Install the WooCommerce Subscriptions plugin in order to create subscriptions.

**For simple subscriptions**

1. Create a simple subscription 
2. Go to the subscription page
3. Click the Payment Request button
4. Displayed items should be correct
5. Payment dialog should include the Shipping Address section

You can use all the cases mentioned below

<details>
<summary>
Simple Subscriptions test list
</summary>

- [x] with Regular price
- [x] with Sale price
- [x] with Tax Status: none
- [x] with Tax Status: taxable, Tax class: Reduced rate 
- [x] with Sign-up fee
- [x] with Sign-up fee + Sale price
- [x] with quantity
- [x] Without stock (button is not displayed)
- [x] With not enough stock (alert with error message should be displayed)
- [x] sold individually

</details>


**For virtual subscriptions**

1. Create a simple subscription 
2. **Check "Virtual" option**
3. Go to the subscription page
4. Click the Payment Request button
5. Displayed items should be correct
6. Payment dialog should **not** include Shipping Address section

You can use all the cases mentioned below

<details>
<summary>
 Simple Subscriptions - Virtual
</summary>

- [x] with Regular price
- [x] with Sale price
- [x] with Tax Status: none
- [x] with Tax Status: taxable, Tax class: Reduced rate 
- [x] with Sign-up fee
- [x] with Sign-up fee + Sale price
- [x] with quantity
- [x] Without stock (button is not displayed)
- [x] With not enough stock (alert with error message should be displayed)
- [x] sold individually

</details>

**For variable subscriptions**

1. Create a variable subscription 
2. Add custom "color" attribute with terms: "red", "green", "blue", with prices, 100 200, 300 respectively. 
3. Create a variation for all attributes
3. Go to the subscription page
4. Select a variation
4. Click the Payment Request button
5. Displayed items should be correct
6. Payment dialog should include the Shipping Address section

You can use all the cases mentioned below

<details>
<summary>Variable Subscription</summary>

- [x] variation with Regular price
- [x] with "Default Form values"
- [x] variation with Sale price
- [x] variation with Tax class: Reduced rate 
- [x] variation with Sign-up fee
- [x] variation with Sign-up fee + Sale price
- [x] variation with quantity
- [x] variation without stock (button is not displayed)
- [x] variation with not enough stock (alert with error message should be displayed)
- [x] variation is virtual

</details>


#### Observations
- This PR does not work with Trial Subscriptions. PR button is hidden for this case.
- PR button shouldn't be displayed if product is added to cart and is set to "Out of stock". 
- PR button shouldn't be displayed if quantity is greater than stock
- Payment dialog asks for Shipping address when selecting a virtual variation
-->

asumaran committed:
Remove duplicate assignment

asumaran committed:
Translate stock validation to the backend

asumaran committed:
Fix error message for variations

asumaran committed:
Remove get_product_price method

asumaran commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -350,7 +332,7 @@ public function get_product_data() {
 		$data['total']           = [
 			'label'   => apply_filters( 'wc_stripe_payment_request_total_label', $this->total_label ),
-			'amount'  => WC_Stripe_Helper::get_stripe_amount( $this->get_product_price( $product ) ),
+			'amount'  => WC_Stripe_Helper::get_stripe_amount( $product->get_price() ),
```
I'm removing `get_product_price` since is not relevant anymore. This amount will never be used in the frontend. Its purpose is just to be able to initialize the PR button when page loads.

asumaran commented on assets/js/stripe-payment-request.js:
```diff
@@ -519,13 +519,8 @@ jQuery( function( $ ) {
 				// First check if product can be added to cart.
-				var addToCartButton = $( '.single_add_to_cart_button' );
-				if ( addToCartButton.is( '.disabled' ) ) {
-					if ( addToCartButton.is( '.wc-variation-is-unavailable' ) ) {
```
I've translated this validation to the backend like the one we do when the user is trying to purchase more products than available. I think it's better to do it in the backend otherwise the user could potentially try to purchase something out of stock based solely on the information of the frontend.

ricardo replied:
That makes sense 👍

asumaran commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1053,34 +1030,40 @@ public function ajax_add_to_cart() {
+					$variation             = wc_get_product( $variation_id );
+					$has_enough_stock      = $variation->has_enough_stock( $quantity );
+					$is_in_stock           = $variation->is_in_stock();
+					$stock_qty_for_display = wc_format_stock_quantity_for_display( $variation->get_stock_quantity(), $variation );
```
Fixes  error message for variations.

asumaran commented:
I'm going to create issues for the items listed in the Observations section once this PR gets approved.

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -803,11 +785,6 @@ private function is_product_supported( $product ) {
 			return false;
 		}
 
-		// Not supported for subscription products when user is not authenticated and account creation is not possible.
-		if ( class_exists( 'WC_Subscriptions_Product' ) && WC_Subscriptions_Product::is_subscription( $product ) && ! is_user_logged_in() && ! $this->is_account_creation_possible() ) {
-			return false;
-		}
-
```
We should not remove this check, otherwise it will throw an error when you try to purchase a subscription without being authenticated when account creation upon checkout is not possible in **WooCommerce > Settings > Accounts and Privacy**.

When porting [WCPay#1811](https://github.com/Automattic/woocommerce-payments/pull/1811) it will fix guest checkout + subscriptions.

asumaran replied:
can you elaborate? I didn't see any error. In fact the PR button is not displayed to the user when authentication is required. Here's the options I tested.

<img width="781" alt="Screen Shot 2021-06-21 at 4 16 59 PM" src="https://user-images.githubusercontent.com/1025173/122829073-1e549680-d2ac-11eb-912b-f11dcb2be51c.png">

asumaran replied:
Apart from that. This PR tries to enable the PR button on the product page so I need to remove the conditional. Maybe I can adjust the conditional but I didn't got any issues by removing it.

ricardo replied:
> In fact the PR button is not displayed to the user when authentication is required.

That's because [we're preventing it](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/cb456e28d1af5eaa272603f60a5ea415aa094d4e/includes/payment-methods/class-wc-stripe-payment-request.php#L735-L737) when both guest checkout + creating an account upon checkout are not possible.

Settings to reproduce:

<img width="806" alt="Screen Shot 2021-06-21 at 20 03 26" src="https://user-images.githubusercontent.com/5509901/122838300-2b817d80-d2cc-11eb-943a-6d0454b850ac.png">

Error:

<img width="827" alt="Screen Shot 2021-06-21 at 20 04 21" src="https://user-images.githubusercontent.com/5509901/122838309-320ff500-d2cc-11eb-962a-2d4e75720c63.png">

> This PR tries to enable the PR button on the product page so I need to remove the conditional.

If you read the last part of the check, it's not about the subscription product, it's about disallowing the PRB for unauthenticated users when account creation is not possible, so we prevent the aforementioned error:

```php
// Not supported for subscription products when user is not authenticated and account creation is not possible.
if ( class_exists( 'WC_Subscriptions_Product' ) && WC_Subscriptions_Product::is_subscription( $product ) && ! is_user_logged_in() && ! $this->is_account_creation_possible() ) {
	return false;
}
```

As you know, we're refactoring how we deal with this error in #1609.

With that said, by reverting this check, maybe we should rename the PR to something related to product variations so it's more appropriate and doesn't get confused when searching. Feel free to ignore.

asumaran replied:
@ricardo  I've tried multiple ways to reproduce the error including setting the options like the image you shared but I get no error. Am I missing something? How do you get the error? When clicking the PR button? What types of subscription do you have? It's a simple subscription or a variable subscription?

ricardo replied:
Are you logged in while testing? You need to be logged out. The error happens when you try to pay using PRB. I tested with a simple subscription, but I think it can be any supported subscription.

asumaran replied:
@ricardo yes. I'm logged out. I've even tried with an incognito window.

asumaran replied:
I've updated this branch with the latest changes from `trunk` and this discussion seems no longer relevant. The PR button now shows an alert when the user is trying to purchase a subscription. 

<img width="693" alt="Screen Shot 2021-06-21 at 11 54 15 PM" src="https://user-images.githubusercontent.com/1025173/122865413-fd139a80-d2eb-11eb-9083-8c50afc35038.png">

ricardo commented on includes/payment-methods/class-wc-stripe-payment-request.php:
```diff
@@ -1053,34 +1030,40 @@ public function ajax_add_to_cart() {
 			}
 
+			if ( ! $is_in_stock ) {
+				throw new Exception( sprintf( __( 'Sorry, this product is unavailable. Please choose a different combination.', 'woocommerce-gateway-stripe' ) ) );
+			}
```
No need to `sprintf` because there's no args in this string.

To avoid inconsistencies with WC core, I think we should add a comment stating where this comes from.

```
// Same error message as in core `wc_add_to_cart_variation_params.i18n_unavailable_text`.
```

asumaran replied:
🤦‍♂️  done fd2e6c0

ricardo reviewed (dismissed):
Tests well 👍 

When removing the check for subscriptions with guest checkout, this PR doesn't really affect subscriptions, so maybe we should rename it to "Payment request button: Improve product variation logic" or something along those lines.

I quickly looked into the trial with subscriptions issue and PR #921, and I'm guessing it's related to the multiple shipping packages PR. If that's the case, I don't think we can offer a proper fix.

The multiple shipping packages issue was recently fixed in #1591 by hiding the PR button.

When you add a shippable subscription with trial to the cart, the shipping options seem to appear as a separate package under "Recurring totals", but the payment request flow does not offer the possibility to select anything from that package.

Actually, if you add a simple subscription with shipping + a simple product with shipping to the cart and try to pay using PRB, it doesn't allow you to select more than one shipping option. Would this also be an issue perhaps?

asumaran committed:
Remove unnecessary sprintf usage

asumaran commented:
> this PR doesn't really affect subscriptions, so maybe we should rename it to "Payment request button: Improve product variation logic" or something along those lines.

@ricardo I agree. Your [PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1609) basically did all of the work for handling Subscriptions. This PR end up being just some minor improvements :)

I've updated the title and removed the testing instructions. It doesn't make sense anymore.

> I quickly looked into the trial with subscriptions issue and PR #921, and I'm guessing it's related to the multiple shipping packages PR

I wasn't looking at that issue when working on this PR.  

> When you add a shippable subscription with trial to the cart...

Trial Subscriptions that require shipping will continue unsupported.

asumaran commented:
Hey @ricardo. This PR end up with some small changes since support for Subscriptions is implemented already. Could you take a look at this? I would like to merge this along with #1602

ricardo reviewed (approved):
LGTM :shipit:
